File: src/main/java/com/d3velop/Renderer/FrameBuffer.java
package com.d3velop.Renderer;

import static org.lwjgl.opengl.GL30.*;

import java.nio.ShortBuffer;

public class FrameBuffer {
    public int id;
    private int fbo;
    private int width;
    private int height;

    FrameBuffer(){
        fbo = glGenFramebuffers();
        glBindFramebuffer(GL_FRAMEBUFFER, fbo);
        id = glGenTextures();
        glBindTexture(GL_TEXTURE_2D, id);
        glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB16F, 0, 0, 0, GL_RGB, GL_UNSIGNED_SHORT, (ShortBuffer) null);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
        glBindTexture(GL_TEXTURE_2D, 0);
        glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, id, 0);
        glBindFramebuffer(GL_FRAMEBUFFER, 0);
    }

    public void size(int width, int height){
        this.width = width;
        this.height = height;
        glBindTexture(GL_TEXTURE_2D, id);
        glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB16, width, height, 0, GL_RGB, GL_UNSIGNED_SHORT, (ShortBuffer) null);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
        glBindTexture(GL_TEXTURE_2D, 0);
        glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, id, 0);
        glBindFramebuffer(GL_FRAMEBUFFER, 0);
    }
    public void bind(){
        glBindFramebuffer(GL_FRAMEBUFFER, fbo);
        glViewport(0, 0, width, height);
        glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
    }
    public void unbind(){
        glBindFramebuffer(GL_FRAMEBUFFER, 0);
    }
}

File: src/framebuffer.cpp
#include "framebuffer.h"

FrameBuffer::FrameBuffer() { this->created = false; };

void FrameBuffer::genFrameBuffer() {
  glGenFramebuffers(1, &fbo);
  glBindFramebuffer(GL_FRAMEBUFFER, fbo);
  glGenTextures(1, &fbid);
  glBindTexture(GL_TEXTURE_2D, fbid);
  glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB16F, 0, 0, 0, GL_RGB, GL_UNSIGNED_SHORT,
               NULL);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
  glBindTexture(GL_TEXTURE_2D, 0);
  glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D,
                         fbid, 0);
  glBindFramebuffer(GL_FRAMEBUFFER, 0);
  this->created = true;
}
void FrameBuffer::size(int width, int height) {
  this->width = width;
  this->height = height;
  glBindTexture(GL_TEXTURE_2D, fbid);
  glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB16, width, height, 0, GL_RGB,
               GL_UNSIGNED_SHORT, NULL);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
  glBindTexture(GL_TEXTURE_2D, 0);
  glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D,
                         fbid, 0);
  glBindFramebuffer(GL_FRAMEBUFFER, 0);
}
void FrameBuffer::bind() {
  glBindFramebuffer(GL_FRAMEBUFFER, fbo);
  glViewport(0, 0, width, height);
  glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
}
void FrameBuffer::unbind() { glBindFramebuffer(GL_FRAMEBUFFER, 0); }
